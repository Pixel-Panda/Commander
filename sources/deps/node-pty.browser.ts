export function spawn() {
    throw new Error('Not implemented in browser environments');
}
