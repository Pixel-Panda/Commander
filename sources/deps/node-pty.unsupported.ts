export function spawn() {
    throw new Error('Not implemented in this environments');
}
