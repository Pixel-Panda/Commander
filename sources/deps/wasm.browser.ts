export const loadMono = async () => require(`mono-layout/wasm`);
export const loadOniguruma = async () => require(`vscode-oniguruma/release/onig.wasm`);
export const loadYogini = async () => require(`yogini/wasm`);
