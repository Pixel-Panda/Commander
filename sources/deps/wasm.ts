import {readFileSync} from 'fs';

export const loadMono = async () => readFileSync(require.resolve(`mono-layout/wasm`));
export const loadOniguruma = async () => readFileSync(require.resolve(`vscode-oniguruma/release/onig.wasm`));
export const loadYogini = async () => readFileSync(require.resolve(`yogini/wasm`));
