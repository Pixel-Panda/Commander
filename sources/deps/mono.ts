import {Context, createContext} from 'mono-layout';

import {loadMono}               from '#sources/deps/wasm';

let mono: Context | undefined;

export async function initMonoLayout() {
  if (typeof mono === `undefined`)
    mono = await createContext(await loadMono());

  return mono;
}

export function createLayout() {
  return mono!.createLayout();
}
