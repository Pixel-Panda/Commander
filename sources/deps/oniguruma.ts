import {loadWASM, OnigScanner, OnigString} from 'vscode-oniguruma';

import {loadOniguruma}                     from '#sources/deps/wasm';

let ready = false;

export async function initOniguruma() {
  if (!ready) {
    await loadWASM(await loadOniguruma());
    ready = true;
  }

  return {
    createOnigScanner,
    createOnigString,
  };
}

export function createOnigScanner(patterns: Array<string>) {
  return new OnigScanner(patterns);
}

export function createOnigString(s: string) {
  return new OnigString(s);
}
