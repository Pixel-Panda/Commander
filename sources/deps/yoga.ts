import {createContext, YogaConfig} from 'yogini';

import {loadYogini}                from '#sources/deps/wasm';

let yoga: YogaConfig | undefined;

export async function initYoga() {
  if (typeof yoga === `undefined`) {
    yoga = await createContext(await loadYogini());
    yoga.setPointScaleFactor(0);
  }

  return yoga;
}

export function createNode() {
  return yoga!.createNode();
}
