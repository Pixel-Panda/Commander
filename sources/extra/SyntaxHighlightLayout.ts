import {TextLayout, TextOperation}                                                     from 'mono-layout';
import {style}                                                                         from 'term-strings';
import {IGrammar, INITIAL, IRawTheme, ITokenizeLineResult2, parseRawGrammar, Registry} from 'vscode-textmate';

import {initOniguruma}                                                                 from '#sources/deps/oniguruma';

export type Theme = {
  tokenColors: IRawTheme[`settings`];
};

export function addSyntaxHighlightSupport(textLayout: TextLayout) {
  const defaultTheme = {tokenColors: []};
  const onRefresh = new Set<() => void>();

  let grammarMap = new Map<string, {scopeName: string}>();

  const onigLib = initOniguruma();
  const makeRegistry = () => new Registry({
    onigLib,
    loadGrammar: async scopeName => {
      const grammar = grammarMap.get(scopeName);
      if (typeof grammar === `undefined`)
        return null;

      return parseRawGrammar(JSON.stringify(grammar), `grammar.json`);
    },
  });

  let activeRegistry: Registry = makeRegistry();
  let activeTheme: Theme = defaultTheme;
  let activeGrammar: IGrammar | null = null;

  let grammarPromise: Promise<void> | null = null;

  const currentTokenizedLines: Array<ITokenizeLineResult2> = [];
  regenerateAll();

  function regenerateAll() {
    regenerateRange({
      startingRow: 0,
      deletedLineCount: 0,
      addedLineCount: textLayout.getRowCount(),
    });

    for (const cb of onRefresh) {
      cb();
    }
  }

  function regenerateRange(op: TextOperation): TextOperation {
    if (!activeGrammar)
      return op;

    const tokenizedLines: Array<ITokenizeLineResult2> = [];

    let startingRow = op.startingRow;

    let deletedLineCount = op.deletedLineCount;
    let addedLineCount = op.addedLineCount;

    while (startingRow > 0 && textLayout.doesSoftWrap(startingRow - 1)) {
      startingRow -= 1;

      deletedLineCount += 1;
      addedLineCount += 1;
    }

    let prevState = startingRow > 0
      ? currentTokenizedLines[startingRow - 1].ruleStack
      : INITIAL;

    for (let t = startingRow, T = t + addedLineCount; t < T; ++t) {
      const tokenizedLine = activeGrammar.tokenizeLine2(realGetLine.call(textLayout, t), prevState);

      tokenizedLines.push(tokenizedLine);
      prevState = tokenizedLine.ruleStack;
    }

    const trailingIndex = startingRow + addedLineCount;
    const trailingLines = textLayout.getRowCount() - trailingIndex;

    for (let t = 0; t < trailingLines; ++t) {
      const tokenizedLine = activeGrammar!.tokenizeLine2(realGetLine.call(textLayout, trailingIndex + t), prevState);

      const previousTokenization = currentTokenizedLines[trailingIndex + t - addedLineCount + deletedLineCount];
      if (tokenizedLine.ruleStack.equals(previousTokenization.ruleStack))
        break;

      tokenizedLines.push(tokenizedLine);
      prevState = tokenizedLine.ruleStack;

      deletedLineCount += 1;
      addedLineCount += 1;
    }

    currentTokenizedLines.splice(startingRow, deletedLineCount, ...tokenizedLines);

    return {
      startingRow,
      deletedLineCount,
      addedLineCount,
    };
  }

  function getColor(metadata: number) {
    const colorMap = activeRegistry.getColorMap();
    const fgColor = colorMap[(metadata & 0b00000000111111111000000000000000) >>> 15];

    return style.color.front(fgColor);
  }

  function getTokenFragment(row: number, start: number, end: number, source: string, tokenIndex: number) {
    const tokens = currentTokenizedLines[row].tokens;

    const startIndex = tokens[tokenIndex];
    const endIndex = tokenIndex + 2 < tokens.length
      ? tokens[tokenIndex + 2]
      : end;

    const fixedStartIndex = Math.max(startIndex, start);
    const fixedEndIndex = Math.min(endIndex, end);

    const slice = source.slice(fixedStartIndex - start, fixedEndIndex - start);
    const metadata = tokens[tokenIndex + 1];

    return getColor(metadata) + slice;
  }

  const realSetSource = textLayout.setSource;
  textLayout.setSource = function (source) {
    return regenerateRange(realSetSource.call(this, source));
  };

  const realSpliceSource = textLayout.spliceSource;
  textLayout.spliceSource = function (start, deleted, added) {
    return regenerateRange(realSpliceSource.call(this, start, deleted, added));
  };

  const realGetLine = textLayout.getLine;
  textLayout.getLine = function (row) {
    const source = realGetLine.call(this, row);
    if (!activeGrammar)
      return source;

    const lineTokens = currentTokenizedLines[row].tokens;

    let result = ``;

    for (let t = 0; t < lineTokens.length; t += 2) {
      const startIndex = lineTokens[t];
      const endIndex = t + 2 < lineTokens.length
        ? lineTokens[t + 2]
        : source.length;

      const slice = source.slice(startIndex, endIndex);
      const metadata = lineTokens[t + 1];

      result += getColor(metadata) + slice;
    }

    return result + style.clear;
  };

  const realGetLineSlice = textLayout.getLineSlice;
  textLayout.getLineSlice = function (row, start, end) {
    const source = realGetLineSlice.call(this, row, start, end);
    if (!activeGrammar)
      return source;

    const lineTokens = currentTokenizedLines[row].tokens;

    let tokenIndex = 0;
    while (tokenIndex + 2 < lineTokens.length && lineTokens[tokenIndex + 2] <= start)
      tokenIndex += 2;

    let result = ``;
    while (tokenIndex < lineTokens.length && lineTokens[tokenIndex] < end) {
      result += getTokenFragment(row, start, end, source, tokenIndex);
      tokenIndex += 2;
    }

    return result + style.clear;
  };

  return {
    onRefresh,

    activate(scopeName: string | null) {
      const loadPromise = scopeName !== null
        ? activeRegistry!.loadGrammar(scopeName)
        : Promise.resolve(null);

      const promise = grammarPromise = loadPromise.then(grammar => {
        if (grammarPromise !== promise)
          return;

        activeGrammar = grammar;
        regenerateAll();
      });

      return promise;
    },

    getGrammars() {
      return [...grammarMap.values()];
    },

    setGrammars(grammars: Array<{scopeName: string}>) {
      const newGrammarMap = new Map();
      for (const grammar of grammars)
        newGrammarMap.set(grammar.scopeName, grammar);

      const currentGrammars = [...grammarMap.values()];
      if (currentGrammars.every(grammar => newGrammarMap.get(grammar.scopeName) === grammar)) {
        grammarMap = newGrammarMap;
        return;
      }

      activeRegistry = makeRegistry();

      regenerateAll();
    },

    getTheme() {
      return activeTheme;
    },

    setTheme(theme: Theme | null) {
      activeTheme = theme ?? defaultTheme;

      activeRegistry?.setTheme({settings: activeTheme.tokenColors});
      regenerateAll();
    },

    getTokens() {
      return currentTokenizedLines.map((tokenizedLine, n) => {
        const source = textLayout.getLine(n);
        const tokens: Array<[string, string, string]> = [];
        const colorMap = activeRegistry.getColorMap();

        for (let t = 0; t < tokenizedLine.tokens.length; t += 2) {
          const startIndex = tokenizedLine.tokens[t];
          const endIndex = t + 2 < tokenizedLine.tokens.length
            ? tokenizedLine.tokens[t + 2]
            : source.length;

          const metadata = tokenizedLine.tokens[t + 1];
          const fgColor = colorMap[(metadata & 0b00000000111111111000000000000000) >>> 15];
          const bgColor = colorMap[(metadata & 0b11111111000000000000000000000000) >>> 24];

          tokens.push([source.slice(startIndex, endIndex), fgColor, bgColor]);
        }

        return tokens;
      });
    },
  };
}
