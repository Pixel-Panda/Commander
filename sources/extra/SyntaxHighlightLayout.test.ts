import theme                          from '#data/themes/WinterIsComing.json';
import {initMonoLayout, createLayout} from '#sources/deps/mono';
import {addSyntaxHighlightSupport}    from '#sources/extra/SyntaxHighlightLayout';

beforeAll(async () => {
  await initMonoLayout();
});

describe(`SyntaxHighlightLayout`, () => {
  describe(`getTokens`, () => {
    it(`supports tokenizing a single line`, async () => {
      const layout = createLayout();
      const syntax = addSyntaxHighlightSupport(layout);

      await syntax.activate(`source.ts`);
      syntax.setTheme(theme);

      layout.setSource(`const foo = 42;\n`);

      expect(syntax.getTokens()).toMatchSnapshot();
    });

    it(`supports expanding a line`, async () => {
      const layout = createLayout();
      const syntax = addSyntaxHighlightSupport(layout);

      await syntax.activate(`source.ts`);
      syntax.setTheme(theme);

      layout.setSource(`const foo = 42;\n`);
      layout.spliceSource(7, 1, `uu: number = 1;\nconst bo`);

      expect(syntax.getTokens()).toMatchSnapshot();
    });

    it(`supports shrinking a line`, async () => {
      const layout = createLayout();
      const syntax = addSyntaxHighlightSupport(layout);

      await syntax.activate(`source.ts`);
      syntax.setTheme(theme);

      layout.setSource(`const foo = 42;\nconst bar = 21;\n`);
      layout.spliceSource(7, 16, `woop`);

      expect(syntax.getTokens()).toMatchSnapshot();
    });
  });

  describe(`getLine`, () => {
    it.only(`supports highlighting a single line`, async () => {
      const layout = createLayout();
      const syntax = addSyntaxHighlightSupport(layout);

      await syntax.activate(`source.ts`);
      syntax.setTheme(theme);

      layout.setSource(`const foo = 42;\n`);

      expect(layout.getLine(0)).toMatchSnapshot();
    });
  });
});
