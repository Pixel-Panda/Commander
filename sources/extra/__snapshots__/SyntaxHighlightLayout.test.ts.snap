// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyntaxHighlightLayout getLine supports highlighting a single line 1`] = `"[38;2;0;191;249m[48;2;255;255;255mconst[38;2;199;146;234m[48;2;255;255;255m [38;2;146;182;244m[48;2;255;255;255mfoo[38;2;199;146;234m[48;2;255;255;255m [38;2;0;191;249m[48;2;255;255;255m=[38;2;199;146;234m[48;2;255;255;255m [38;2;141;236;149m[48;2;255;255;255m42[38;2;188;231;255m[48;2;255;255;255m;[m"`;
